### DISCUSSION TITLE
v2.0 proposed breaking changes

### DISCUSSION THREAD

**Discussion by elalish** [google/model-viewer#3460] [Ideas]
We're coming up on the 2-year mark since `<model-viewer>` 1.0, having done 12 releases since then. We have been carefully following semantic versioning rules to avoid breaking changes for our diverse user base, but this has caused parts of our API to become less beginner-friendly as our newer best-practices have not become defaults. This makes it feel like the time has come to reset with a v2.0 release.

To be clear, we still want to make updating to v2.0 easy for our users, so we're mostly looking to change features that are little-used or will improve the experience on most sites. We can't promise when v2.0 will ship, but hopefully sometime this summer. This thread is to gather feedback on our proposal; please help us make v2.0 our best release yet!

The following is a roughly-prioritized list of breaking changes we're considering. This list will be updated as the conversation progresses. Much of this is designed to reduce cruft in our code base to ensure we minimize bugs and keep code quality high. Please comment regarding any that you agree or disagree with and why. Suggestions of other changes that you would like to see are also welcome. 

 - Minimum radius is doubled to match the [`enable-pan`](https://modelviewer.dev/docs/#entrydocs-stagingandcameras-attributes-enablePan) default, with default field-of-view narrowed to 30 degrees (12 degree default minimum). This gives the default framing a "longer lens" with reduced perspective warp.
 - Default `enable-pan` to true, replacing the `enable-pan` attribute with `disable-pan`. 
 - Remove the ['bounds'](https://modelviewer.dev/docs/#entrydocs-stagingandcameras-attributes-bounds) attribute, making the behavior equivalent to `bounds="tight"`. 
 - Remove the deprecated [`scene-graph-ready`](https://modelviewer.dev/docs/#entrydocs-scenegraph-events-sceneGraphReady) event.
 - Remove the [`progress-mask`](https://modelviewer.dev/docs/#entrydocs-loading-parts-defaultProgressMask); equivalent to setting its part to `display: none;`. This will keep the poster from appearing lightened during loading.
 - Remove the "interaction" option from the [`reveal`](https://modelviewer.dev/docs/#entrydocs-loading-attributes-reveal) attribute. Use "manual" and `dismissPoster()` instead to achieve this behavior.
 - Remove the [`interaction-policy`](https://modelviewer.dev/docs/#entrydocs-stagingandcameras-attributes-interactionPolicy) attribute, allowing only the default "always-allow" behavior. 
 - Remove the "when-focused" option from the [`interaction-prompt`](https://modelviewer.dev/docs/#entrydocs-stagingandcameras-attributes-interactionPrompt) attribute.
 - Remove the [`interaction-prompt-style`](https://modelviewer.dev/docs/#entrydocs-stagingandcameras-attributes-interactionPromptStyle) attribute, allowing only the default "wiggle" behavior. 
 - Remove [`--poster-color`](https://modelviewer.dev/docs/#entrydocs-loading-css-posterColor), making the poster always have a transparent background.
 - Remove the [`seamless-poster`](https://modelviewer.dev/docs/#entrydocs-loading-attributes-seamlessPoster) attribute, making the behavior as though this option is always enabled. Note this in combination with the previous bullet means there is no longer a transition fade between poster and model, so seamless posters as generated by [`toBlob({idealAspect: true})`](https://modelviewer.dev/docs/#entrydocs-loading-methods-toBlob) or our [editor](https://modelviewer.dev/editor/) are strongly recommended. 
 - Switch [`touch-action`](https://modelviewer.dev/docs/#entrydocs-stagingandcameras-attributes-touch-action) default from `pan-y` to `none`. 
 - Switch default [`environment-image`](https://modelviewer.dev/docs/index.html#entrydocs-lightingandenv-attributes-environmentImage) to `neutral`, renaming the current default as `legacy`. 
 - Switch our `package.json` `module` to point to `lib` instead of `dist`. For JS users, `main` will still point to our bundle in `dist`, but this change will send TS users to our lib so they can build and tree-shake it themselves, which is best practice. Also, our lib already contains all our `*.d.ts` files.
 - Remove the `Image.setURI` method from our materials API. Use the [`createTexture`](https://modelviewer.dev/docs/index.html#entrydocs-scenegraph-methods-createTexture) method instead.
 - Update `exportScene` options to match three.js r143. Specifically, remove `embedImages` (always true now) and `forcePowerOfTwoTextures`. Add `forceIndices`.
 - Remove `preload` event, as there is no longer any distinction from the `load` event.
 - Enable USDZ autogeneration by default, by switching the `ar-modes` default to `webxr scene-viewer quick-look`.

**Comment by elalish** [comment]
FYI (a very incomplete list of some users who may have opinions on this): @pushmatrix @mikkoh @mrdoob @donmccurdy @hybridherbst @lucadalli

**Comment by mikkoh** [comment]
I'll need to think about this in great detail. However:

> Minimum radius is doubled to match the [enable-pan](https://modelviewer.dev/docs/#entrydocs-stagingandcameras-attributes-enablePan) default, with default field-of-view narrowed to 30 degrees (12 degree default minimum). This gives the default framing a "longer lens" with reduced perspective warp.

I love an FOV of 30 degrees. The reduced perspective warp is way better for commerce applications.

> Remove the [seamless-poster](https://modelviewer.dev/docs/#entrydocs-loading-attributes-seamlessPoster) attribute, making the behavior as though this option is always enabled. Note this in combination with the previous bullet means there is no longer a transition fade between poster and model, so seamless posters as generated by [toBlob({idealAspect: true})](https://modelviewer.dev/docs/#entrydocs-loading-methods-toBlob) or our [editor](https://modelviewer.dev/editor/) are strongly recommended.

This one is a bit hairy for us as we'd need to backfill and generate all preview images again but it might be worth it. 🤔 We need to experiment with this and just do some thinking around this.

If we're thinking of a 2.0. One thing I'd like to propose is componentizing Model Viewer more. I think one of the best things about Model Viewer is how quickly you can get up and running and have a consistent API for users/developers to interact with models. Increased adoption for 3D. One of the downsides I see of Model Viewer is that there's less innovation in rendering technology. I'd love it if the webcomponent was separated from the renderer. I've done some prototyping around this where I had a webcomponent wrap <model-viewer> and it acts as a translation layer between <model-viewer> or a custom renderer. This might be a fine approach but not ideal.

Also I'd like to add that having a consistent renderer across multiple channels is great. Like making sure 3d models are always rendered the same regardless of where you distribute content is really good. But I wonder if there should be some leeway here for increased innovation.

Pinging
@sleroux and @tylerrowsell on this also.

**Reply by elalish** [reply]
Agreed that separating MV into a core web component bundle and a renderer bundle would be nice. It'll be quite a bit of work, and we don't have anyone pushing hard on this just now, so I don't think that'll make v2.0. However, I believe that can be done as a non-breaking change anyway, so it can come along when it's ready.

**Comment by donmccurdy** [comment]
> Remove [--poster-color](https://modelviewer.dev/docs/#entrydocs-loading-css-posterColor), making the poster always have a transparent background.

+1, I got tripped up on this option at one point.

***

What would you think of changing the units of the `exposure` property to "stops" / EVs? I think three.js may be in the minority of graphics tools in using a linear multiplier there. Possibly applies to `KHR_image_formation` proposal as well.

```js
renderer.toneMappingExposure = Math.pow(2, modelViewerExposure);
```

**Reply by elalish** [reply]
I dunno; agreed that might have been nicer to start with, but at this point it seems like a pretty minor ergonomics issue to justify a breaking change.

**Comment by 3d-game-studio** [comment]
Looking for Best 3d Models Pack Collection?
Check out here- https://www.deep3dsea.com/

deep3dsea.com is a 3d model marketplace for buy & sell 3d models. Choose any Category & Get 100 3d models for only $300 from here- https://www.deep3dsea.com/seller-aaa-game-studio/

here you can find the best 3d models & inspect with 3d viewer. 
Thanks & Regards

**Reply by clochardM33** [reply]
"Looking for Best 3d Models Pack Collection? "No.

**Comment by clochardM33** [comment]
Here are the proposed items sorted from my world view.

**Unaffected by - covered manually and varies by model**
Minimum radius 

**Unaffected by - force to this already**
Default enable-pan to true, replacing the enable-pan attribute with disable-pan. So long as can still add my override code that updates some onscreen camera info overlays when panning and scrolling
Remove [--poster-color]

**Unaffected by** 
Remove the ['bounds']
Remove the deprecated [scene-graph-ready]
Remove the [progress-mask]
Remove the [seamless-poster]

**Didn't know it was there**
Remove the "when-focused" option 
Remove the [interaction-policy]

**Don't care**
Switch [touch-action]
Remove the "interaction" option from the [reveal]
Remove the [interaction-prompt-style] provided that you can still set interaction-prompt="none"

**Comment by mikkoh** [comment]
I've been thinking about the 30 degree FOV as default. I still like this FOV better but I may not been clear above that we'd also need to regenerate all preview images for all of our merchant models

**Reply by elalish** [reply]
Yes, no doubt, hence it being a breaking change. Obviously you can set your FoV attribute to get back to the old style if you haven't had time to update them yet. But I figured better to rip the bandaid off with all the breaking changes at once, so at least you only have to regenerate once.

**Comment by hybridherbst** [comment]
One thing that I'd love to see in 2.0 but not sure if you want to go that route: being able to attach custom GLTFLoader extensions more easily. Note that this wouldn't need to be "official" or "supported", just "possible" would be enough – e.g. one more symbol or so that allows doing more advanced things without having to recompile / rebuild model-viewer from scratch.

**Reply by elalish** [reply]
I'm certainly open to this, and it doesn't sound like a breaking change anyway, so it would be fine in e.g. v2.1 as well. Would you like to open a PR so we can discuss?

**Comment by hybridherbst** [comment]
I had missed this one:

> Remove the "interaction" option from the [reveal](https://modelviewer.dev/docs/#entrydocs-loading-attributes-reveal) attribute. Use "manual" and dismissPoster() instead to achieve this behavior.

which unfortunately means that the behaviour of "click on image to get 3D model" now requires a programmer - not ideal... (noticed because it broke some of our "model galleries" and now custom code is required instead of just writing HTML).

The only example for that is rolled into the "Customize loading experience with your own HTML" example, which adds additional complexity. A separate example for just "click on image to get 3D" would be great.

**Reply by elalish** [reply]
Related, we also had someone recently lament the removal of the "basic" interaction prompt style. I'm happy to revisit. Part of why I removed "interaction" was that it felt kind of broken: it only worked for a click, while most mobile users were expecting a swipe to also make it load. Plus we weren't testing it very well. 

PRs certainly welcome, especially for examples! I achieved my v2.0 goal of vastly simplifying our internal loading logic; if we need to add a few features back in, I can be convinced.
